internal/server/service.go

```diff
@@ -417,16 +418,19 @@ func (s *Service) createMiddleware(options ServiceOptions, certManager CertManag
 func (s *Service) serviceRequestWithTarget(w http.ResponseWriter, r *http.Request) {
 	LoggingRequestContext(r).Service = s.name
 
-	if s.shouldRedirectToHTTPS(r) {
-		s.redirectToHTTPS(w, r)
-		return
-	}
-
-	if !s.options.TLSEnabled && r.TLS != nil {
+    // If TLS is not enabled for this service, reject TLS requests immediately
+    if !s.options.TLSEnabled && r.TLS != nil {
 		SetErrorResponse(w, r, http.StatusServiceUnavailable, nil)
 		return
 	}
 
+    // Perform canonical host and/or TLS redirects in a single hop when needed
+    if url := s.redirectURLIfNeeded(r); url != "" {
+        w.Header().Set("Connection", "close")
+        http.Redirect(w, r, url, http.StatusMovedPermanently)
+        return
+    }
```

Comment:
I'd extract this to its own method, similar to how `handlePausedAndStoppedRequests` looks below. There's enough going on in the details to make it worthwhile, and you can avoid the need for a comment since the method name will be clear.